File: src/BaseHmiTypes/Converters/AutomationXml/HmiTextListToAutomationXmlConverter.cs
using System.Globalization;
using System.Xml;
using BaseHmiTypes.TextGraphicLists;

namespace BaseHmiTypes.Converters.AutomationXml;

public static class HmiTextListToAutomationXmlConverter
{
    public static void WriteAutomationXml(XmlWriter writer, HmiTextList textList, ref int id)
    {
        writer.WriteStartElement("Hmi.TextGraphicList.TextList");
        writer.WriteAttributeString("ID", (id++).ToString(CultureInfo.InvariantCulture));
        writer.WriteStartElement("AttributeList");
        WriteElementString(writer, "ListRange", textList.RangeType.ToString());
        WriteElementString(writer, "Name", textList.Name);
        writer.WriteEndElement();

        if (textList.Comment != null)
        {
            writer.WriteStartElement("ObjectList");
            HmiAutomationXmlConverter.WriteAutomationXml(textList.Comment, writer, ref id, "Comment");
            writer.WriteEndElement();
        }

        foreach (var entry in textList.Entries)
        {
            writer.WriteStartElement("Hmi.TextGraphicList.TextListEntry");
            writer.WriteAttributeString("ID", (id++).ToString(CultureInfo.InvariantCulture));
            writer.WriteAttributeString("CompositionName", "Entries");
            writer.WriteStartElement("AttributeList");
            WriteElementString(writer, "DefaultEntry", entry.Default);
            WriteElementString(writer, "From", entry.From);
            WriteElementString(writer, "To", entry.To);
            writer.WriteEndElement();

            if (entry.Text != null)
            {
                writer.WriteStartElement("ObjectList");
                HmiAutomationXmlConverter.WriteAutomationXml(entry.Text, writer, ref id, "Items");
                writer.WriteEndElement();
            }

            writer.WriteEndElement();
        }

        writer.WriteEndElement();
    }

    private static void WriteElementString(XmlWriter writer, string name, object? value)
    {
        writer.WriteElementString(name, value switch
        {
            null => string.Empty,
            bool boolean => boolean ? "true" : "false",
            IFormattable formattable => formattable.ToString(null, CultureInfo.InvariantCulture),
            _ => value.ToString()
        });
    }
}

File: typescript/base-hmi-types/src/converters/automation-xml/HmiTextListToAutomationXmlConverter.ts
import { HmiTextList } from "../../text-graphic-lists/HmiTextList.js";
import { HmiAutomationXmlConverter } from "./HmiAutomationXmlConverter.js";
import { XmlWriter } from "./XmlWriter.js";

export class HmiTextListToAutomationXmlConverter {
  static writeAutomationXml(writer: XmlWriter, textList: HmiTextList, id: { value: number }): void {
    writer.writeStartElement("Hmi.TextGraphicList.TextList");
    writer.writeAttributeString("ID", String(id.value++));
    writer.writeStartElement("AttributeList");
    writer.writeElementString("ListRange", textList.rangeType);
    writer.writeElementString("Name", textList.name);
    writer.writeEndElement("AttributeList");

    if (textList.comment !== undefined) {
      writer.writeStartElement("ObjectList");
      HmiAutomationXmlConverter.writeMultilingualText(textList.comment, writer, id, "Comment");
      writer.writeEndElement("ObjectList");
    }

    for (const entry of textList.entries) {
      writer.writeStartElement("Hmi.TextGraphicList.TextListEntry");
      writer.writeAttributeString("ID", String(id.value++));
      writer.writeAttributeString("CompositionName", "Entries");
      writer.writeStartElement("AttributeList");
      writer.writeElementString("DefaultEntry", entry.default);
      writer.writeElementString("From", entry.from);
      writer.writeElementString("To", entry.to);
      writer.writeEndElement("AttributeList");
      if (entry.text !== undefined) {
        writer.writeStartElement("ObjectList");
        HmiAutomationXmlConverter.writeMultilingualText(entry.text, writer, id, "Items");
        writer.writeEndElement("ObjectList");
      }
      writer.writeEndElement("Hmi.TextGraphicList.TextListEntry");
    }

    writer.writeEndElement("Hmi.TextGraphicList.TextList");
  }
}
